### DISCUSSION TITLE
[apex] isTestMethodOrClass and triggers

### DISCUSSION THREAD

**Discussion by adangel** [pmd/pmd#3059] [Q&A]
We use the following code to throw violations where a test class should use test methods. isTestMethodOrClass(node) is the predefined method used here to check whether a node here is test method or class.

it is working fine for test classes where the violations are being captured correctly.

Unfortunately this violation is flagging for triggers also where as it should only be flagged for test classes. Please find the snippet of the violation for a trigger in the attachment.

Code starts here:

```java
public Object visit(final ASTMethod node, final Object data) {
    // test methods should have @isTest annotation.
    if (isTestMethodOrClass(node)) {
        return data;
    }
    return checkForAssertStatements(node, data);
}

private Object checkForAssertStatements(final ASTMethod testMethod, final Object data) {
    List<astmethodcallexpression> methodCallList = testMethod.findDescendantsOfType(ASTMethodCallExpression.class);
    String assertMethodName;
    for (ASTMethodCallExpression assertMethodCall : methodCallList) {
        assertMethodName = assertMethodCall.getFullMethodName().toLowerCase(Locale.ROOT);
        if ((!testMethod.getModifiers().isTestOrTestSetup() && (ASSERT_METHODS.contains(assertMethodName))) || (!testMethod.getModifiers().isTestOrTestSetup())) {
            addViolationWithMessage(data, testMethod,
                "Use @isTest or TestMethod or @TestSetup annotation for methods in Test Class.",
                new Object[] { testMethod.getImage() });
            return data;
        }
    }
    return data;</astmethodcallexpression>
```

![trigger](https://user-images.githubusercontent.com/77477451/104684375-ea82ad00-571e-11eb-9467-3c20dd72a393.png)

_Originally posted by @aklmmap in https://github.com/pmd/pmd/discussions/2998#discussioncomment-283546_

**Comment by adangel** [comment]
@aklmmap Thanks for the question. I have some follow up questions for clarification:

* You seem to use a custom rule. Is this correct?
* Can you describe what your rule should do?

From the snippet I see, it seems like it is flagging assert method calls, when they occur outside of test classes. Is that correct?

Then I don't understand, why the ContactTrigger is flagged, because I don't see a assert method call...

I very good tool to analyze the AST, e.g. the ContactTrigger, and develop rules is the designer -> see https://pmd.github.io/latest/pmd_userdocs_extending_designer_reference.html

In order to copy the code snippet into the Designer, I've typed it here from what I could read from the screenshot:

```apex
trigger ContactTrigger on Contact (before insert, before update, before delete, after insert, after update, after delete) {
    ContactTriggerDAO contactDAO = new ContactTriggerDAO();
    if (contactDAO.getTriggerFlag('Contact')) {
        new ContactTriggerHandler().run();
    }
}
```

I can only see `getTriggerFlag` and `run` as MethodCallExpressions. They don't look like asserts...

**Reply by rsoesemann** [reply]
@rmohan20 FYI

**Reply by aklmmap** [reply]
This is a custom apex rule we have developed. 
Rule: Testclass or methods should have assert statements and @testsetup methods. 

The issue with this code was it was firing for all triggers. 

Fortunately we have found a solution for that too. Thank you for your attention and support.

**Reply by adangel** [reply]
would you mind to share the solution? then we can mark it as an answer....

**Reply by aklmmap** [reply]
We also checked the parent of the ASTMethod node, and checked the type of its node. For triggers it would return the type "UserTrigger". Hence we have added this condition in the code where it would flag violations for triggers.
